src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java

```diff
@@ -22,86 +22,85 @@
 
 package jdk.crac.impl;
 
-import jdk.crac.CheckpointException;
-import jdk.crac.Context;
-import jdk.crac.Resource;
-import jdk.crac.RestoreException;
-import sun.security.action.GetBooleanAction;
+import jdk.crac.*;
+import jdk.internal.crac.LoggerContainer;
 
-import java.util.*;
-import java.util.stream.Collectors;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
-public abstract class AbstractContextImpl<R extends Resource, P> extends Context<R> {
+public abstract class AbstractContextImpl<R extends Resource> extends Context<R> {
+    protected List<Resource> restoreQ = null;
 
-    private static class FlagsHolder {
-        public static final boolean DEBUG =
-            GetBooleanAction.privilegedGetProperty("jdk.crac.debug");
+    protected static <E extends Exception> void recordExceptions(E source) {
+        Throwable[] suppressed = source.getSuppressed();
+        if (suppressed.length == 0) {
+            Core.recordException(source);
+        }
+        for (Throwable t : suppressed) {
+            Core.recordException(t);
+        }
     }
 
-    private WeakHashMap<R, P> checkpointQ = new WeakHashMap<>();
-    private List<R> restoreQ = null;
-    private Comparator<Map.Entry<R, P>> comparator;
+    protected void setModified(R resource, String msg) {
+        Core.recordException(new CheckpointException(
+                "Adding resource " + resource + " to " + this + (msg != null ? msg : "")));
+    }
 
-    protected AbstractContextImpl(Comparator<Map.Entry<R, P>> comparator) {
-        this.comparator = comparator;
+    protected void invokeBeforeCheckpoint(Resource resource) {
+        LoggerContainer.debug("beforeCheckpoint {0}", resource);
+        // Resource.afterRestore is invoked even if Resource.beforeCheckpoint fails
+        restoreQ.add(resource);
+        try {
+            resource.beforeCheckpoint(semanticContext());
+        } catch (CheckpointException e) {
+            recordExceptions(e);
+        } catch (Exception e) {
+            Core.recordException(e);
```

Comment:
The contexts are supposed to run all resources regardless of whether any failed, therefore there's no point in propagating and re-wrapping the exceptions. We could have a method to check whether this C/R is 'marked for rollback' (has any exceptions), but I don't see why would the parent context decide on the type - because it would get just CheckpointException with a list of other failures from components it does not understand.

About the distinction: the difference is that if you get a CheckpointException you'd unwrap it, recording only the inner suppressed ones. But I should push that to `recordExceptions` and rather decide based on CheckpointException message than number of suppressions.

Reply:
> The contexts are supposed to run all resources regardless of whether any failed, therefore there's no point in propagating and re-wrapping the exceptions. ...  I don't see why would the parent context decide on the type - because it would get just CheckpointException with a list of other failures from components it does not understand.

The parent Context may implement an artbitrary handling (for example, unloading a component compleltely if that is throwing an exception). So throwing an exception is something useful.

With that, the new Core.recordException is completely new exception flow, that just opimizes somthing the generic throw scheme. With that, the generic schemes should be something good enough already, we don't need to complicate the interface, the code,.. CheckpointExcepotions are still exceptions, that is, we don't expect them often, there is no need to optimize them.

Reply:
It's not about optimization (in the sense of performance) but about removing code bloat, and the need for parent context to tediously copy failures (deciding whether something was a wrapper exception or the actual failure), when all you need to do in the end is to report them in bulk. It's not a new exception flow, it's removing the flow as there is no exception flow needed.

Your example is invalid: If a throwing resource is to be removed, it's the task of the parent context - and that one will see the exception. The parent context should not remove its child context since one of the N resources in the child context is failing.

Reply:
> You're trading additional code for a complicated interface. What are the directions what to use: Core.recordException() vs throw new Exception() ? This two very similar interfaces are the sign we are trying to do something strange.

That's because of the non-standard requirement to continue after encountering an exception, and multiple rewraps in a hierarchy of contexts.
Resources should normally throw exceptions; Contexts are here to call into the Resource and aggregate errors. There's no point of propagating error higher up the hierarchy.

> Could you elaborate?.. I'm lost what "remove child context" means. I was reffering to the parent context being able to handle a CheckpointException from the child context.

And why would the child context throw the CheckpointException? The most common reason would be that one of X resources in child context has thrown. It's fine if a (child) Context removes the inner Resource after throwing. But that error should not propagate any higher, because if parent context were to remove it's (throwing) child context it would remove it along with the other X - 1 resources that are working correctly.

Reply:
I realized I haven't stressed enough one motivation for `Core.registerException`: when the context has finished its `beforeCheckpoint` (it won't be called again) and someone calls into this Context's `register` we cannot propagate the failure through throwing. So we won't avoid API change - and in my view it's natural to use the same method for reporting the failure during `beforeCheckpoint` execution, too.